compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp

```diff
@@ -470,6 +471,125 @@ struct ReverseOpConversion : public OpConversionPattern<mhlo::ReverseOp> {
   }
 };
 
+//===----------------------------------------------------------------------===//
+// TopkOp
+//===----------------------------------------------------------------------===//
+
+struct TopkOpConversion : public OpConversionPattern<chlo::TopKOp> {
+  using OpConversionPattern<chlo::TopKOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      chlo::TopKOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    Location loc = op.getLoc();
+    Value operand = adaptor.operand();
+
+    auto inputValuesType = operand.getType().dyn_cast<RankedTensorType>();
+    if (!inputValuesType) {
+      return failure();
+    }
+    uint64_t kDim = inputValuesType.getRank() - 1;
+
+    // Create a range of indicies for the input tensor match input shape but
+    // with i32 type.
+    SmallVector<Value> dynSizes;
+    for (auto en : llvm::enumerate(inputValuesType.getShape())) {
+      if (en.value() == ShapedType::kDynamicSize) {
+        dynSizes.push_back(rewriter.create<tensor::DimOp>(
+            loc, adaptor.getOperands()[0], en.index()));
+      }
+    }
+    Value initAcc = rewriter.create<linalg::InitTensorOp>(
+        loc, dynSizes, inputValuesType.getShape(), rewriter.getI32Type());
+
+    // Create the indexing maps for the generic
+    size_t inputRank = inputValuesType.getRank();
+    AffineMap mapIdentity = rewriter.getMultiDimIdentityMap(inputRank);
+    SmallVector<AffineMap> indexingMaps{mapIdentity};
+    SmallVector<StringRef> iterators{inputRank, "parallel"};
+    Value inputIndices =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, /*resultTensorTypes=*/initAcc.getType(),
+                /*inputs=*/ValueRange{}, /*outputs=*/ValueRange{initAcc},
+                indexingMaps, iterators,
+                [=](OpBuilder &b, Location loc, ValueRange args) {
+                  // Indicies increase along the k dimension
+                  Value index = b.create<linalg::IndexOp>(loc, kDim);
+                  Value intIndex =
+                      b.create<arith::IndexCastOp>(loc, b.getI32Type(), index);
+                  b.create<linalg::YieldOp>(loc, intIndex);
+                })
+            .getResult(0);
+
+    // Create and initialize output tensors for LinalgExt TopK results
+    // Define the output types based on the results of CHLO TopK
+    auto outputValuesType = op.getResultTypes()[0].dyn_cast<ShapedType>();
+    auto outputIndicesType = op.getResultTypes()[1].dyn_cast<ShapedType>();
```

Comment:
Prefer using readable methods from adaptor.

E.g., `adaptor.values().getType().cast...`, `adaptor.indices()...`.

Reply:
These methods don't exist on the `TopKOpAdaptor`:

```
class TopKOpAdaptor {
public:
  TopKOpAdaptor(::mlir::ValueRange values, ::mlir::DictionaryAttr attrs = nullptr, ::mlir::RegionRange regions = {});

  TopKOpAdaptor(TopKOp op);

  ::mlir::ValueRange getOperands();
  std::pair<unsigned, unsigned> getODSOperandIndexAndLength(unsigned index);
  ::mlir::ValueRange getODSOperands(unsigned index);
  ::mlir::Value operand();
  ::mlir::DictionaryAttr getAttributes();
  ::mlir::IntegerAttr kAttr();
  uint64_t k();
  ::mlir::LogicalResult verify(::mlir::Location loc);
private:
  ::mlir::ValueRange odsOperands;
  ::mlir::DictionaryAttr odsAttrs;
  ::mlir::RegionRange odsRegions;
  ::llvm::Optional<::mlir::OperationName> odsOpName;
};
```

Reply:
ah, it's reasonable because adaptor is more for accessing operands. I think these methods exist on the `TopKOp`. We're able to use `op.values().getType().cast...` etc.